dex/runner.go

```diff
@@ -0,0 +1,40 @@
+package dex
+
+import (
+	"context"
+	"sync"
+)
+
+type Runner interface {
+	Run(ctx context.Context) //error
+}
+
+type StartStopWaiter struct {
+	runner Runner
+	wg     sync.WaitGroup
+	cancel context.CancelFunc
+	ctx    context.Context
+}
```

Comment:
Some potential benefits to the `type dex.Runner interface { Run(context.Context) }` pattern and the implemented `StartStopWaiter` type, listed to convince myself as well:

- Kicking off subsystem goroutines explicitly (not from the constructor) feels like the best reason to me.  However, this can be done with Start/Stop/Wait too.  We're less likely to mismanage goroutines this way as we're forced to create them ourselves, with `Run` called inside.
- A consistent interface (`dex.Runner`) describing how subsystems should be written is also appealing.  It allows multiple subsystems to be managed in a slice (or other container) as I've done.  It also sets an example pattern for future subsystems to use.
- With `Run(ctx)` the subsystem is stopped by cancelling the context, or any parent context.  The `StartStopWaiter` wraps a provided context so that subsystems can be stopped separately, but a less graceful emergency shutdown is still possible if the parent context is canceled, say by a deferred cancel function, which would be executed in the case of a panic whereas `DEX.Done()` would not necessarily be depending on how it is called or if itself crashes for some reason.